pkg/serializer/internal/stream/json_payload_builder.go

```diff
@@ -103,6 +104,61 @@ const (
 func (b *JSONPayloadBuilder) BuildWithOnErrItemTooBigPolicy(
 	m marshaler.IterableStreamJSONMarshaler,
 	policy OnErrItemTooBigPolicy) (forwarder.Payloads, error) {
+	var receiver rawPayloadsReceiver
+	err := b.buildWithOnErrItemTooBigPolicy(
+		m,
+		policy,
+		&receiver)
+
+	return receiver.payloads, err
+}
+
+// payloadsReceiver receives payloads created by buildWithOnErrItemTooBigPolicy
+type payloadsReceiver interface {
+	onItemAddedToNextPayload(m marshaler.IterableStreamJSONMarshaler)
+	append(payload *[]byte)
+}
+
+// rawPayloadsReceiver is a simple receiver that stores payload as []*[]byte
+type rawPayloadsReceiver struct{ payloads []*[]byte }
+
+func (p *rawPayloadsReceiver) append(payload *[]byte) { p.payloads = append(p.payloads, payload) }
+
+func (p *rawPayloadsReceiver) onItemAddedToNextPayload(m marshaler.IterableStreamJSONMarshaler) {}
+
+// BuildWithOnErrItemTooBigPolicyMetadata serializes a metadata payload and sends it to the forwarder
+func (b *JSONPayloadBuilder) BuildWithOnErrItemTooBigPolicyMetadata(
+	m marshaler.IterableStreamWithPointJSONMarshaler,
+	policy OnErrItemTooBigPolicy) (transaction.BytesPayloads, error) {
+	var receiver metadataPayloadsReceiver
+	err := b.buildWithOnErrItemTooBigPolicy(
+		m,
+		policy,
+		&receiver)
+
+	return receiver.payloads, err
+}
+
+// metadataPayloadsReceiver is a receiver that stores payloads as BytesPayloads
+type metadataPayloadsReceiver struct {
+	payloads   transaction.BytesPayloads
+	pointCount int
+}
+
+func (p *metadataPayloadsReceiver) append(payload *[]byte) {
+	p.payloads = append(p.payloads, transaction.NewBytesPayload(payload, p.pointCount))
+	p.pointCount = 0
+}
+
+func (p *metadataPayloadsReceiver) onItemAddedToNextPayload(m marshaler.IterableStreamJSONMarshaler) {
+	p.pointCount += m.(marshaler.IterableStreamWithPointJSONMarshaler).GetCurrentItemPointCount()
```

Comment:
Would it be possible to extend `IterableStreamJSONMarshaler` so that it always produces point count (possibly always zero, if it is not needed/supported)? I think that would obviate the need for the type cast here and the second receiver type.

Or at least there should be a comment that explains that `metadataPayloadsReceiver` can only be used with a certain marshaller.

Reply:
As you probably noticed, there are two kind of payloads `transaction.BytesPayloads` and `forwarder.Payloads`. These payloads are produced by the serializers and consumed by the forwarder. I did my best to make it explicit when a serializer has metadata and when it doesn't have any. I could remove `IterableStreamJSONMarshaler` and replace by `IterableStreamWithPointJSONMarshaler` but that will make the code less explicit regarding if a serializer create metadata or not.

`metadataPayloadsReceiver` is a local type that is designed to be used at a single location. Anyway, I have added a comment on `metadataPayloadsReceiver` to make it more explicit as you suggested.

Reply:
To be clear, I don't think this issue is a blocker, but on a larger scale, I don't understand why the split between `transaction.BytesPayloads` and `forwarder.Payloads` is needed: both are handled exactly the same in the end (`DefaultForwarder#appendPayloads` converts `forwarder.Payloads` to `transaction.BytesPayloads` erasing the difference). Hypothetically, if we were to add telemetry about dropped service checks and events, would we add more payload types to highlight different contents?

Reply:
If each serializer returns `transaction.BytesPayloads`, then if you want to know if a particular reader handles metadata or not the metadata, you have to read the code of the serializer. In this PR, I have tried to make it explicit using different types (`transaction.BytesPayloads` when the serializer returns metadata and `forwarder.Payloads` when it doesn't).

One of the issue with the current implementation of `buildWithOnErrItemTooBigPolicy` is that it doesn't allow easily to have different behavior for different kind of metric (series or events for example). It makes the code of this PR more verbose as it should be.

From the description of the PR:
`When using series API V1 and when JSON stream is not enabled, the number of dropped points is not reported (this is a deprecated mode). By default, JSON stream is enabled and series API V2 will replaced soon API V1.` This is also a good reason to have both types as it makes very explicit which series serializer support metadata  in the code.

> Hypothetically, if we were to add telemetry about dropped service checks and events, would we add more payload types to highlight different contents?

If different serializers returned different data, then it would make sense that each serializer use a different type.